### ISSUE TITLE
indentation of buildInputs is very odd

### ISSUE THREAD

**Issue by peterhoeg** [nix-community/nixpkgs-fmt#137]
Not sure what the proper way to do this would be, but the nested indentation is not right.

## Input
```nix
{ mkDerivation }:

mkDerivation rec {

  buildInputs = [
    alsaLib
  ]  ++ lib.optionals stdenv.isLinux [
     utillinux
  ] ++ (with gst_all_1; [
     gstreamer
  ]);
}

```

## Output
```nix
{ mkDerivation }:

mkDerivation rec {

  buildInputs = [
    alsaLib
  ]
    ++ lib.optionals stdenv.isLinux [
         utillinux
       ]
    ++ (
         with gst_all_1; [
           gstreamer
         ]
       )
    ;
}

```

## Desired output

```nix
{ mkDerivation }:

mkDerivation rec {

  buildInputs = [
    alsaLib
  ]
  ++ lib.optionals stdenv.isLinux [
    utillinux
  ]
  ++ (with gst_all_1; [
    gstreamer
  ])
  ;
}
```

**Comment by matklad** [comment]
Seems similar to #123. We currently have a rule that forces all bin ops to the new line, to make diffing easy.

**Comment by peterhoeg** [comment]
I've updated the original post with how I think it should look - and this wouldn't change anything from a diff point of view. My issue is the increasing indentation.

**Comment by matklad** [comment]
>I've updated the original post with how I think it should look - and this wouldn't change anything from a diff point of view. My issue is the increasing indentation.

Ah, than that would be https://github.com/nix-community/nixpkgs-fmt/issues/132#issuecomment-525245531. 

One of the  motivations for indent is to reduce the diff if the code is refactored to 

```
  buildInputs = [ alsaLib ]
    ++ lib.optionals stdenv.isLinux [
         utillinux
       ]
    ++ (
         with gst_all_1; [
           gstreamer
         ]
       )
    ;
```

or to

```
  buildInputs = 
    [ alsaLib ]
    ++ lib.optionals stdenv.isLinux [
         utillinux
       ]
    ++ (
         with gst_all_1; [
           gstreamer
         ]
       )
    ;
```

In general, implementation wise this `++` pattern seems like a special case.

But I also see how the current formatting is not aesthetically ideal.

@zimbatm so, what should we do here?

* Try to convince everybody that this formatting is better?
* revert #135 in favor of just fixing #132 directly?
* brainstorm some third option for formatting, which is both regular and beautiful?

**Comment by zimbatm** [comment]
Sorry for the back and forth.

The current implementation is breaking the (arbitrary) rule of not more than 1 indent diff between lines.

How about this.

1. Change the indent based on the first line. If the line ends in `=\s*\n$` after the key, indent +1. Otherwise keep the same indent as the key.
2. Remove the `++` rule.

In that case, all of these would be valid:

```nix
buildInputs = [
  foo
];
```

If the user wants to break it up:
```nix
buildInputs =
  [ foo ];
```

Then it's up to the user to choose if they want `++` on their own line:

```nix
buildInputs = [ foo ]
++ lib.optional stdenv.isLinux [
  utillinux
];
```

```nix
buildInputs = [
  foo
] ++ lib.optional stdenv.isLinux [
   utillinux
];
```

```nix
buildInputs =
  [
    foo
  ]
  ++ lib.optional stdenv.isLinux [
    utillinux
  ]
  ;
```

In that case, the only rule that forces the `\n` on the first line is a function attrset argument that is on multiple lines:
```nix
myFun = # <-- \n is forced here
  { a
  , b
  }:
  a + b;
```

Basically we're making a decision of not making a decision, just focus on the indent.

**Comment by matklad** [comment]
@zimbatm gave it a shot in https://github.com/nix-community/nixpkgs-fmt/pull/139/. I think being more lax about linebreaks is a good thing in this case. Unlike with lists or function arguments, where we have a genuine list of items, for binary operations we have an associative structure where the elements could be arbitrary, and messing with it has a higher change of causing problems.
